assets/course-theme/sidebar.js

```diff
@@ -0,0 +1,249 @@
+
+	// If the sidebar is very tall and does not fit into the viewport vertically
+	// we scroll the sticky sidebar up until the bottom is reached. Or we scroll
+	// the sticky sidebar down until the top of the sidebar is reached.
+	if ( stickySidebarIsTallerThanViewport && ! initialPosition ) {
+		stickySidebarNewTop = stickySidebarRect.top - delta;
+		const stickySidebarNewBottom = stickySidebarRect.bottom - delta;
+		const stickySidebarMinTop = sidebarRect.top;
+		const stickySidebarMinBottom =
+			window.innerHeight - SIDEBAR_BOTTOM_MARGIN;
+
+		// The sidebar is moving upwards.
+		if ( delta >= 0 ) {
+			if ( stickySidebarNewBottom < stickySidebarMinBottom ) {
+				stickySidebarNewTop =
+					stickySidebarMinBottom - stickySidebarHeight;
+			}
+
+			// The sidebar is moving downwards.
+		} else {
+			if ( stickySidebarNewTop > headerRect.bottom ) {
+				stickySidebarNewTop = headerRect.bottom;
+			}
+			if ( stickySidebarNewTop < stickySidebarMinTop ) {
+				stickySidebarNewTop = stickySidebarMinTop;
+			}
+		}
+
+		// If the sidebar fits into the viewport vertically
+		// then we simply stick it below the header when user
+		// scrolls it up above the header.
+	} else if ( sidebarRect.top <= headerRect.bottom ) {
+		stickySidebarNewTop = headerRect.bottom;
+
+		// By default we position the sticky sidebar on top
+		// of the original sidebar.
+	} else {
+		stickySidebarNewTop = sidebarRect.top;
+	}
+
+	// Need to subtract the sidebar top margin because fixed positioned elements
+	// are pushed down by css top margin.
+	stickySidebar.style.top = `${ stickySidebarNewTop - sidebarMarginTop }px`;
+}
+
+/**
+ * Reinitializes the sticky sideber
+ */
+const reinitializeSidebar = debounce( () => {
+	preparestickySidebar();
+	updateSidebarPosition( true );
+}, 500 );
+
+/**
+ * Makes sure the height of the sidebar is at least the height
+ * of the featured video in 'modern' LM template.
+ */
+function syncSidebarSizeWithVideo() {
+	if ( featuredVideo && sidebar ) {
+		new window.ResizeObserver( () => {
+			const videoHeight = featuredVideo.offsetHeight;
+			const sidebarHeight = sidebar.offsetHeight;
+			if (
+				! videoHeight ||
+				! sidebarHeight ||
+				sidebarHeight >= videoHeight
+			) {
+				return;
+			}
+			sidebar.style.height = `${ videoHeight }px`;
+			reinitializeSidebar();
+		} ).observe( featuredVideo );
+	}
+}
+
+/**
+ * Makes the sidebar sticky for relevant LM templates.
+ */
+function setupStickySidebar() {
+	if ( ! isStickySidebar() ) {
+		return;
+	}
+
+	queryDomElements();
+
+	document.defaultView.addEventListener( 'scroll', () =>
+		updateSidebarPosition()
+	);
+
+	// eslint-disable-next-line @wordpress/no-global-event-listener
+	window.addEventListener( 'resize', reinitializeSidebar );
+
+	// Make sure sidebar height is not shorter than the video height
+	// for `moderm` lm template.
+	if ( document.body.classList.contains( 'learning-mode--modern' ) ) {
+		syncSidebarSizeWithVideo();
+	}
```

Comment:
I think we can tie this one the Lesson Video block, it's really a feature of that block. In `featured-video-size.js` it actually already has a resize observer and height shenanigans going.

Reply:
> In featured-video-size.js it actually already has a resize observer and height shenanigans going.

Yep. In fact I used most of the logic from that file in here.

> I think we can tie this one the Lesson Video block, it's really a feature of that block.

Not sure if I agree. It might sound like this logic belongs into the `featured-video-size.js` file, because it is already dealing with the sizing of the video block and we might as well add the sizing for sidebar in there. Especially if we want the height of the sidebar be synced with the video block.

But I think the logic that deals with the size of the sidebar block should belong to the rest of the logic that deals with sidebar business. A developer, looking at the `sidebar.js` and not being familiar with the `featured-video-size.js` will not be able to guess that there is additional stuff that are related to sidebar in the `featured-video-size.js`.

Also, if in the future, we need to keep the sidebar height in sync with another block that is not video or related to video. Let's say we will introduce a slideshow block. Just like the featured video, the slideshow would be a lesson slideshow and will be rendered at top of the lesson content. In that case we could keep adding the sizing logic of the sidebar in the `sidebar.js`.